server/src/main/java/org/opensearch/index/mapper/FlatObjectFieldMapper.java

```diff
@@ -0,0 +1,745 @@
+import org.opensearch.search.aggregations.support.CoreValuesSourceType;
+import org.opensearch.search.lookup.SearchLookup;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Supplier;
+
+import static org.opensearch.search.SearchService.ALLOW_EXPENSIVE_QUERIES;
+
+/**
+ * A field mapper for flat_objects. This mapper accepts JSON object and treat as string fields in one index.
+ * A flat_object field contains one parent field itself and two substring fields:
+ * field._valueAndPath and field._value
+ * @opensearch.internal
+ */
+public final class FlatObjectFieldMapper extends DynamicKeyFieldMapper {
+
+    public static final String CONTENT_TYPE = "flat_object";
+    private static final String VALUE_AND_PATH_SUFFIX = "._valueAndPath";
+    private static final String VALUE_SUFFIX = "._value";
+    private static final String DOT_SYMBOL = ".";
+    private static final String EQUAL_SYMBOL = "=";
+
+    /**
+     * In flat_object field mapper, field type is similar to keyword field type
+     * Cannot be tokenized, can OmitNorms, and can setIndexOption.
+     * @opensearch.internal
+     */
+    public static class Defaults {
+        public static final FieldType FIELD_TYPE = new FieldType();
+
+        static {
+            FIELD_TYPE.setTokenized(false);
+            FIELD_TYPE.setOmitNorms(true);
+            FIELD_TYPE.setIndexOptions(IndexOptions.DOCS);
+            FIELD_TYPE.freeze();
+        }
+
+    }
+
+    @Override
+    public MappedFieldType keyedFieldType(String key) {
+        return new FlatObjectFieldType(this.name() + DOT_SYMBOL + key);
+    }
+
+    /**
+     * FlatObjectFieldType is the parent field type.
+     */
+    public static class FlatObjectField extends Field {
+
+        public FlatObjectField(String field, BytesRef term, FieldType ft) {
+            super(field, term, ft);
+        }
+
+    }
+
+    private static FlatObjectFieldMapper toType(FieldMapper in) {
+        return (FlatObjectFieldMapper) in;
+    }
+
+    /**
+     * The builder for the flat_object field mapper using default parameters as
+     * indexed: flat_object field mapper is default to be indexed.
+     * hasDocValues: to store index and support efficient access to individual field values.
+     * stored: the original value of the field is not stored in the index.
+     * nullValue: not accept null value
+     * ignoreAbove: exclude values that exceed the maximum length from the indexing process.
+     * @opensearch.internal
+     */
+    public static class Builder extends FieldMapper.Builder<Builder> {
+
+        private final IndexAnalyzers indexAnalyzers;
+
+        Builder(String name, IndexAnalyzers indexAnalyzers) {
+            super(name, Defaults.FIELD_TYPE);
+            builder = this;
```

Comment:
what's the use of this line?

Reply:
it can be used later on if we want to open up parameters for analyzers, such as synonyms.

Reply:
then use it in the next PR https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it